Userland/Libraries/LibCore/Process.cpp

```diff
@@ -163,6 +167,35 @@ ErrorOr<bool> Process::is_being_debugged()
         return (tracer_pid != 0UL);
     }
     return false;
+#elif defined(AK_OS_GNU_HURD)
+    process_t proc = getproc();
+    if (!MACH_PORT_VALID(proc))
+        return Error::from_syscall("getproc"sv, -errno);
+
+    int flags = PI_FETCH_TASKINFO;
+    struct procinfo pi_buffer;
+    int* pi_array = reinterpret_cast<int*>(&pi_buffer);
+    mach_msg_type_number_t pi_array_len = sizeof(pi_buffer) / sizeof(int);
+    data_t waits = nullptr;
+    mach_msg_type_number_t waits_len = 0;
+
+    kern_return_t err = proc_getprocinfo(proc, getpid(), &flags, &pi_array, &pi_array_len, &waits, &waits_len);
+    mach_port_deallocate(mach_task_self(), proc);
+    if (err) {
+        __hurd_fail(static_cast<error_t>(err));
+        return Error::from_syscall("proc_getprocinfo"sv, -errno);
+    }
+
+    VERIFY(pi_array_len >= sizeof(struct procinfo));
+    struct procinfo* procinfo = reinterpret_cast<struct procinfo*>(pi_array);
```

Comment:
This seems unnecessary, can't we just use pi_buffer directly? Or does the code you were referencing for this implementation have the pi_buffer be an array of ints that's the size of a procinfo?

Reply:
That's how all Mach APIs/RPCs work: when they have an `out` array parameter, they have you pass in a buffer (as a `char**` or some such) and its size (as a `mach_msg_type_number_t*`). Then it will either write over the buffer you gave it (and update the size), or (typically in case things don't fit, but also possibly in other cases) allocate a different buffer, and return its base ptr / size to you. You should then do
```c
if (returned_buffer != &my_buffer)
    munmap(returned_buffer, returned_size);
```
once you're done with the data.

> Or does the code you were referencing for this implementation have the pi_buffer be an array of ints that's the size of a procinfo?

Yes. MIG doesn't have a proper concept of structures (GNU MIG does since very recently), so whenever a structure has to be passed over IPC it's just passed as an array of integers.

Also this specific API can return info for more than just a single task, depending on the flags (like it can also give you info about all the threads), though we're not using that here. This is another reason why the buffer size is variable in the API.

Reply:
ah I see. so we're doing this b/c the kernel might allocate us a buffer? fair enough, though I would rather my kernel give me an error with the correct size than give me a malloc'd buffer :)

Reply:
Another reason for passing external buffers is that Mach tries to use zero-copy techniques by playing with virtual memory whenever it makes sense. Like if task A sends task B a large suitably aligned memory buffer over IPC, Mach won't actually memcpy it byte by byte, it'd just make the same memory pages appear in the address space of task B, using copy-on-write to make this transparent. So this is also why APIs like this one can return a different buffer than the one you have passed into it.

And then of course Unix happens, and Unix insists that e.g. `read()` must fill the provided buffer and not give you a more convenient one, which if you think of it makes little sense, but such is the API. So glibc literally had to memcpy the read data back into the original buffer if it did arrive out of line via copy-on-write transfer. Apple's / OSF Mach version at least has `mach_msg_overwrite()` for cases like these; but GNU Mach does not 🙁

Reply:
Added some explanations about this to the code, please check if it is clear enough.

Reply:
Yep, I saw that on my glance over, thanks for the explanation! I think it's a good enough explainer to be a good reference when people ask "how does GNU/Hurd IPC/RPC work from the caller side" in the future